### PR TITLE
logger: Add support for logging only on bat power

### PR DESCRIPTION
**Pull request by tzai**
Will not start logging from boot if battery is not connected.

**Problem**
For the logging modes that cause logger to start at boot (boot until disarm or boot until shutdown), a log is started every time the FCU is powered. This includes connecting to the FCU over USB for firmware updates, parameter modifications, or other non-flight modifications. This causes the creation of logs in scenarios where they are most likely unneeded, which clutters up the log space.

**Solution**
Introduce a parameter (`SDLOG_BOOT_BAT`) that when enabled, prevents logs from being started from boot if a battery is not connected.

**Note**
Only works for the first battery instance. If FCU has second power connector that is used instead, this will not work as-is.

### CODE REVIEW

**Comment by dagar** [comment]
Good idea, maybe even worth enabling by default.

**Comment by dagar** [comment]
> **Note**
> Only works for the first battery instance. If FCU has second power connector that is used instead, this will not work as-is.

Given that you're only checking once at logger start another option could be vehicle_status_flags.condition_battery_healthy. https://github.com/PX4/Firmware/blob/89693258a9deed3a7d61b8c398f7c951da6f6b48/msg/vehicle_status_flags.msg#L16

File: src/modules/logger/params.c [hunk 1]
```diff
@@ -66,6 +66,18 @@ PARAM_DEFINE_INT32(SDLOG_UTC_OFFSET, 0);
  */
 PARAM_DEFINE_INT32(SDLOG_MODE, 0);
 
+/**
+ * Battery-only Logging
+ *
+ * When enabled, logging will not start from boot if battery power is not detected
+ * (e.g. powered via USB on a test bench). This prevents extraneous flight logs from
+ * being created during bench testing.
+ *
```
Comment (mhkabir) [inline src/modules/logger/params.c]: Add another line:
Note that this applies only to log-from-boot modes. It has no effect in arming-based modes.

**Comment by mhkabir** [comment]
> Given that you're only checking once at logger start another option could be vehicle_status_flags.condition_battery_healthy.

**Comment by mhkabir** [comment]
> Given that you're only checking once at logger start another option could be vehicle_status_flags.condition_battery_healthy.

I like this one. We may also check for `condition_battery_healthy && !usb_connected` from the same message, and that way it won't log even if on battery but connected to a USB. (assuming this is something people would want)

**Comment by tzai** [comment]
> > **Note**
> > Only works for the first battery instance. If FCU has second power connector that is used instead, this will not work as-is.
> 
> Given that you're only checking once at logger start another option could be vehicle_status_flags.condition_battery_healthy.
> 
> https://github.com/PX4/Firmware/blob/89693258a9deed3a7d61b8c398f7c951da6f6b48/msg/vehicle_status_flags.msg#L16

This would work, but it might have unintended consequences. If the FCU reboots with any battery warning level other than none, it won't start a log. Also, I just tested on the Pixhawk 4, and `condition_battery_healthy` does not get set to false at first, even though `battery_status.battery_warning` is `BATTERY_WARNING_EMERGENCY`. The battery warning was simulated by connecting a 3S battery while configured for 6S. Using `condition_battery_healthy` would effectively accomplish the same thing.

File: src/modules/logger/params.c (hunk 1 above)
Comment (tzai) [reply src/modules/logger/params.c]: Thanks, clarified.

File: src/modules/logger/logger.cpp [hunk 2]
```diff
@@ -518,6 +520,29 @@ void Logger::run()
 {
 	PX4_INFO("logger started (mode=%s)", configured_backend_mode());
 
+	bool boot_logging_bat_only = false;
```
Comment (bkueng) [inline src/modules/logger/logger.cpp]: Can you move the code below into a separate method, so that you get:
```
const bool disable_boot_logging = getDisableBootLogging();
```

And move that down directly before it's used?

File: src/modules/logger/logger.cpp [hunk 3]
```diff
@@ -518,6 +520,29 @@ void Logger::run()
...
+	bool disable_boot_logging = false;
+
+	if (_boot_bat_only != PARAM_INVALID) {
+		param_get(_boot_bat_only, &boot_logging_bat_only);
+	}
+
+	if (boot_logging_bat_only) {
+		uORB::Subscription battery_status_sub{ORB_ID(battery_status)};
+
+		if (battery_status_sub.updated()) {
```
Comment (bkueng) [inline src/modules/logger/logger.cpp]:
```suggestion
		if (battery_status_sub.copy(&battery_status)) {
```

File: src/modules/logger/logger.cpp [hunk 4]
```diff
@@ -518,6 +520,29 @@ void Logger::run()
 {
 	PX4_INFO("logger started (mode=%s)", configured_backend_mode());
 
+	bool boot_logging_bat_only = false;
+	bool disable_boot_logging = false;
+
+	if (_boot_bat_only != PARAM_INVALID) {
+		param_get(_boot_bat_only, &boot_logging_bat_only);
```
Comment (bkueng) [inline src/modules/logger/logger.cpp]: `boot_logging_bat_only` needs to be `int32_t`

@dagar why did you change that in https://github.com/PX4/Firmware/commit/060463e8a7436701394f0ebe2d4ce568c7780289#diff-49dacc90dcc99405772dbb3540d7665fL485? It would have been caught.

File: src/modules/logger/logger.cpp (hunk 2 above)
Comment (tzai) [reply src/modules/logger/logger.cpp]: Alright, will do

**Comment by tzai** [comment]
@bkueng thanks for the review, I implemented your suggestions.
